Code/GraphMol/CIPLabeler/CIPLabeler.cpp

```diff
@@ -0,0 +1,193 @@
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <algorithm>
+#include <memory>
+
+#include <GraphMol/RDKitBase.h>
+
+#include "CIPLabeler.h"
+#include "CIPMol.h"
+#include "configs/Sp2Bond.h"
+#include "configs/Tetrahedral.h"
+
+#include "rules/Rules.hpp"
+#include "rules/Rule1a.h"
+#include "rules/Rule1b.h"
+#include "rules/Rule2.h"
+#include "rules/Rule3.h"
+#include "rules/Rule4a.h"
+#include "rules/Rule4b.h"
+#include "rules/Rule4c.h"
+#include "rules/Rule5New.h"
+#include "rules/Rule6.h"
+
+namespace RDKit {
+namespace CIPLabeler {
+
+namespace {
+
+// constitutional rules
+const Rules constitutional_rules({new Rule1a, new Rule1b, new Rule2});
+
+// all rules (require aux calc)
+const Rules all_rules({new Rule1a, new Rule1b, new Rule2, new Rule3, new Rule4a,
+                       new Rule4b, new Rule4c, new Rule5New, new Rule6});
+
+std::vector<std::unique_ptr<Configuration>> findConfigs(const CIPMol &mol) {
+  std::vector<std::unique_ptr<Configuration>> configs;
+
+  for (auto &atom : mol.atoms()) {
+    auto chiraltag = atom->getChiralTag();
+    if (chiraltag == Atom::CHI_TETRAHEDRAL_CW ||
+        chiraltag == Atom::CHI_TETRAHEDRAL_CCW) {
+      std::vector<Atom *> carriers;
+      carriers.reserve(4);
+      for (auto &bond : mol.getBonds(atom)) {
+        auto nbr = bond->getOtherAtom(atom);
+        carriers.push_back(nbr);
+      }
+      if (carriers.size() < 4) {
+        // Implicit H -- use the central atom instead of a dummy H
+        carriers.push_back(atom);
+      }
+
+      auto atom_cfg = -1;
+      if (chiraltag == Atom::CHI_TETRAHEDRAL_CW) {
+        atom_cfg = Tetrahedral::RIGHT;
+      } else {
+        atom_cfg = Tetrahedral::LEFT;
+      }
+
+      std::unique_ptr<Tetrahedral> cfg{
+          new Tetrahedral(mol, atom, std::move(carriers), atom_cfg)};
+      configs.push_back(std::move(cfg));
+    }
+  }
+
+  for (auto i = 0u; i < mol.getNumBonds(); ++i) {
+    auto bond = mol.getBond(i);
+
+    int bond_cfg = -1;
+    switch (bond->getStereo()) {
+    case Bond::STEREOE:
+    case Bond::STEREOTRANS:
+      bond_cfg = Sp2Bond::OPPOSITE;
+      break;
+    case Bond::STEREOZ:
+    case Bond::STEREOCIS:
+      bond_cfg = Sp2Bond::TOGETHER;
+      break;
+    default:
+      continue;
+    }
+
+    auto stereo_atoms = MolOps::findStereoAtoms(bond);
+    std::vector<Atom *> anchors{
+        {mol.getAtom(stereo_atoms[0]), mol.getAtom(stereo_atoms[1])}};
+    std::vector<Atom *> atoms{{bond->getBeginAtom(), bond->getEndAtom()}};
+
+    std::unique_ptr<Sp2Bond> cfg(
+        new Sp2Bond(mol, bond, std::move(atoms), std::move(anchors), bond_cfg));
+    configs.push_back(std::move(cfg));
+  }
+
+  return configs;
+}
+
+bool labelAux(const std::vector<std::unique_ptr<Configuration>> &configs,
```

Comment:
Aren't you modifying the `Configuration`s pointed to by some of the elements of `configs`?
If so, I think it's misleading/wrong to have it be `const`

Reply:
Good catch. It used to be correct until I made the configurations own the `Digraph`, since this function doesn't change the configuration, except for building/expanding the `Digraph`.